---
title: "Validating Happie's Phage Detection Scheme"
date: "last update: `r format(Sys.Date(), format = '%d %B %Y')`" 
author: "Nicholas Waters"
output:
  tufte::tufte_html: 
    tufte_features: [ "italics"]
  tufte::tufte_handout: default
  pdf_document: default
  html_document: default
editor_options:
  chunk_output_type: console
toc: true
toc_float:
  collapsed: false
  smooth_scroll: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("tufte")
# setwd("~/GitHub/happie/validation/")
library(tidyverse)

#source("http://bioconductor.org/biocLite.R")
#biocLite("IRanges")
library(IRanges)
```

# Introduction:
In order to determine the appropriateness of ProphET for detecting prophage content in the environmental (and clinical) *E. coli*, we will compare the results of our pipeline ("happie") to the supplementary data  from Dr. Sherwood Casjens' paper "Prophages and bacterial genomics: what have we learned so far?"^[ <https://doi.org/10.1046/j.1365-2958.2003.03580.x>].

Happie uses ProphET as its primary phage detection tool, after comparing with the other state-of-the-art tools in the analysis described previously^[ <https://nickp60.github.io/weird_one_offs/testing_3_prophage_finders/>].  Additionally, happie uses IslandPAth-DIMOB to detect genomic islands, which sometimes (but not always) are phage-related.  Here, we descibe how the results from happie compare with the results from Casjens 2003.

# Data

All the data and scripts can be found at <https://github.com/nickp60/happie/>, under the *validation* directory.

We downloaded the word doc of the supplementary data table from Casjens 2013, *mmi_3580_sm_tables1.doc* (why are tables in a .doc) and then transformated that into a slightly less terrible excel file *casjen2003.xlsx*[sic]. This new table we added a link to the genome on NCBI, and an FTP path to download the genome. This was difficult, as between 2003 and now, everything from gene names to taxonomy changed.  The original link to the genomes repository at Venter is broken, and so I had to spend a fair amount of time manually combing the databases for likely strains. Of additional help is the following document, from PhageFinder: <http://phage-finder.sourceforge.net/302_bact_gen.pdf>. Only 1 of the entries I was unable to locate,  due to not finding the genome itself, or the locus tags having changed so much that no comparison was possible.

 The missing genome: Xanthomonas axonopodis 903.


```{marginfigure}
Due to a git mishap, I lost the version of casjen2003.xlsx that had ftp paths; the paths were there when (below) I created the ~casjen2003_newloci.xlsx~ file, so those can be consulted instead.
```


<!-- ```{r, eval=FALSE} -->


<!-- dat <- readxl::read_xlsx("./casjen2003.xlsx", sheet = 2) -->
<!-- dat <- dat %>% fill(host, assembly) #%>% View() -->

<!-- ``` -->



To see if the listed locus tags corresponded with the locus tags in the genbank file, I downloaded one of these genomes:

```{r, eval=FALSE}
dir.create("genomes")
test_file_path <- "genomes/tmp.gbk"
this_path <- "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/009/065/GCF_000009065.1_ASM906v1/GCF_000009065.1_ASM906v1_genomic.gbff.gz"
if (!file.exists(test_file_path)){
  cmd <- paste("curl -L", this_path, "| gunzip  >", test_file_path)
  print(cmd)
  system(cmd)
}
```

Now, for this genome, we need to see if a locus tag given ("AgrC0835"") actually match whats in the file.  They dont in this case^[old locus tag is "Agr_C_835"]. The fun continues: some had different cases, issues with leading 0's, missing locus tags, and more. A quick back-of-the-envelope assessment hinted that the time spent trying to parse would be too high; these must be added manually.


We need a new copy of the table to annotate, and after making this file, I added FTP paths for all the assemblies.

```{r, eval=T}
file.copy(from = "./casjen2003.xlsx", to="./casjen2003_newloci.xlsx", overwrite = F)
# sheet 2 I mannually added any prophet hits that were absent in original data
new_dat <- readxl::read_xlsx("casjen2003_newloci.xlsx", sheet = 3)
# fill down columns
new_dat <- new_dat %>% fill(host, ftp, assembly, sequence, n) 

# simplify name
new_dat$short_name <- gsub("\\W+", "_", new_dat$host )
# seqname is used to compare later to GFFs
new_dat$seqname <-  gsub("(.*)\\..*", "\\1", new_dat$sequence)
# create a path to the genbank file from the fasta file
new_dat$ftp_gbk <- gsub(".fna.gz", ".gbff.gz", new_dat$ftp)

```


I downloaded all the genomes to a folder called `genomes`:
```{r eval=F}
for (i in 1:nrow(new_dat)){
  test_file_path <- file.path("genomes", paste0(new_dat$short_name[i], ".gbk"))
  this_path <- new_dat[i, "ftp_gbk"]
  if (!file.exists(test_file_path)){
    cmd <- paste("curl -L", this_path, "| gunzip  >", test_file_path)
    print(cmd)
    system(cmd)
  }
}
``` 


Next I had the enviable jot of going entry by entry, and trying to find the annotated phage region's coords now that the locus tags are different.  In practice, this meant finding the first and last listed locus tags (or the next one, if the first could not be found), and entering in the start or end coord, respectively.

```{marginfigure}
Occasionally (such as for E. coli EDL933) I would have to get the RefSeq (GCA) version of the Genbank (GCF) version, as the locus tags failed to match.
```


Escherichia coli O157:H7 EDL933 appeared to be using locus tag nomencalture from another organisms, so I had to skip that one.  Genes that were from the same operon had locus tags scattered in distant parts of this genome, in contrast to what was reported in Casjens 2003.

I was unable to match annotations for the following organisms:

- Escherichia coli O157:H7 EDL933
- Mesorhizobium loti MAFF303099

In some cases, the excel file was modified to fix naming issues: for instance, I found that the dash in Shewanella MRâ€“1 needs to be changed to a UTF-8 hyphen, to avoid errors with prokka and the em-dashes.

In instances of just single genes annotated, I just took the first one's start and end coords; we will ignore these downstream, as the table mentions every integrase, even if not in a phagey region.

In the meantime, I ran happie on all those genomes, and prettied up the results, combining them into a single results file:

<!-- counter=0;for i in ./genomes/*.gbk; do echo $counter $i; name=$(basename ${i%.*}); if [ -d "${name}_casjens" ] ; then echo "skipping ${name}"; else happie --contigs $i -o ${name}_casjens ; fi ; counter=$((counter+1)); done -->

```{sh, eval=FALSE}
counter=0; 
for i in ./genomes/*.gbk; 
do 
  echo $counter $i;
  name=$(basename ${i%.*}); 
  if [ -d "${name}_casjens" ] ; 
  then 
    echo "skipping ${name}"; 
  else 
    happie --contigs $i -o ${name}_casjens ; 
  fi ; 
  counter=$((counter+1)); 
done
cat ./*/mobile_genome_coords >> combined_happie_coords

```


# Evaluating Results
I had to do a bit of munging file names to make sure all the names line up. 

<!-- After another round of time-estimation, I decided to score hits by hand, with the following criteria. -->
Previous verison of this script replied on hand-assessed overlaps in a "status" column, but this was converted to IRanges assessment.There was extremely good agreement between the hand-scored results and the machine-scored results, but as there were a few missed by hand and detected with IRanges, I used that exclusivbely from here on out, with the following criteria.

- A hit would be considered valid of the majority of the sequence overlapped.  This is a fuzzy (and probably generous) estimation, but neccessary as the paper reports everything they found, from single genes to two intact and neighboring prophages.In some cases^[see Ralstonia_solanacearum_GMI1000, NC_003295.1,879632-93943], the prophet hit was entirely contained within the reported hit;  keeping a hard scoring threshold would result in accurately reporting the detection, so we changed the length threshold to a more generous definition: greater than half the length of $min(subject\_end-subject\_start, query\_end-query\_start)$.  
- See the docs^[<https://www.rdocumentation.org/packages/IRanges/versions/2.0.1/topics/findOverlaps-methods> ] for definition of overlaps.  I used the "any" type.


We read in the combined results file.  As happie does not report entries for those lacking hit, we manually created dummy rows to add the results table:
```{bash, eval=FALSE}
### get those that had empty results files to make dummy entries
for i in ./*/; do if [ ! -s "${i}mobile_genome_coords" ]; then echo $i; fi done
```

```{r}
happie_dat_raw <- read.csv("combined_happie_coords", sep="\t", header = F, col.names =c("path", "program", "type", "sequence", "start", "stop"), stringsAsFactors = F)

# These lacked phage hits in Casjens 2003
missing_include <- c(
  "Buchnera_aphidicola_Ap_casjens/",
  "Buchnera_aphidicola_Bp_casjens/",
  "Buchnera_aphidicola_Sg_casjens/",
  "Mycoplasma_genitalium_G_37_casjens/",
  "Mycoplasma_pneumoniae_M129_casjens/")
# these were missing data; see above. either the annotations were missing or the genome itself was
missing_exclude <- c(
  "Escherichia_coli_O157_VT-2_Sakai_casjens/",
  "Mesorhizobium_loti_____MAFF303099",
  "Xanthomonas_axonopodis__903_casjens/")
# add missing directory
tmp <- data.frame(
  path=paste0("/home/1/2/3/4/", missing_include),  
  program=rep(NA, length(missing_include)),
  type=rep(NA, length(missing_include)),
  sequence=rep(NA, length(missing_include)),
  start=rep(NA, length(missing_include)),
  stop=rep(NA, length(missing_include))
)
happie_dat <- rbind(happie_dat_raw, tmp)
happie_dat$short_name <- gsub("\\/(.*?)\\/(.*?)\\/(.*?)\\/(.*?)\\/(.*?)\\/(.*?)\\/(.*)", "\\6", happie_dat$path)
happie_dat$short_name <- gsub("_casjens", "", happie_dat$short_name)

# I cant seem to nest by program name, so I have to filter
prophet_nested = happie_dat %>% filter(program=="prophet") %>% group_by(short_name, sequence) %>% nest(.key="prophet", start, stop)
dimob_nested = happie_dat %>% filter(program=="dimob") %>% group_by(short_name, sequence)%>% nest(.key="dimob", start, stop)

nested_happie <- full_join(prophet_nested, dimob_nested, by=c("short_name", "sequence"))

# These is our check to see if the merge will be successful, that the names will be matching
new_dat$short_name[!(new_dat$short_name %in% happie_dat$short_name)]
happie_dat$short_name[!(happie_dat$short_name %in% new_dat$short_name)]
```


## Genes per Loci
The criteria used by ProphET, Phage_Finder, etc included a filtering step to remove any loci of less than 7 CDS's from the Casjens table.  To do this, we first needed to convert the GenBank files to GFF, for use with readseq^[<http://iubio.bio.indiana.edu/cgi-bin/readseq.cgi>]:

```{bash, eval=FALSE}
# gbk to gff conversion
conda install readseq
mkdir genomes_gffs;for i in genomes/*.gbk; do bn=$(basename "$i"); filename="${bn%.*}" ; echo $filename; readseq -f 24 $i -o genomes_gffs/${filename}.gff ; done

```
And then reading in the GFF files, in one of the least efficient ways possible. 

```{r, echo=TRUE, results="hide"}
new_dat$n_genes <- NA
nrows <-nrow(new_dat)
for (i in 1:nrow(new_dat)){
  print(paste(i, "of", nrows))
  path = file.path("genomes_gffs", paste0(new_dat$short_name[i], ".gff")) 
  this_gff <- read.csv2(
    path,  comment.char = "#", header=F, sep="\t", 
    col.names = c("seqname",	"source",	"feature",	"start",	"end",	
                  "score",	"strand",	"frame",	"attributes")
  ) %>% filter(feature=="CDS")

  new_dat$n_genes[i] <- this_gff %>% 
    filter(seqname == new_dat$seqname[i]) %>% 
    filter(start >= new_dat$valid_locus_start[i] & 
             end <= new_dat$valid_locus_end[i]) %>%
    nrow()
}

new_dat$enough_genes <- new_dat$n_genes >= 7
```

Here we can see that only `r table(new_dat$enough_genes)[[2]]` of the `r nrow(new_dat)` had loci of at least 7 CDSs.

## Discrepancies between reports

The data contrained in the `casjens_neewloci.xlsx` shows a total of `r new_dat %>% nrow()` 251 entries, from `r length(unique(new_dat$short_name))` organisms. `r new_dat %>% filter(enough_genes) %>% nrow()`, after removing those with CDSs having fewer than 7 genes (from now `r nrow(unique(new_dat %>% filter(enough_genes) %>% select(short_name)))` organisms). ACLAME (prophinder) claims to used the same data, although they list 54 organisms, and say that there are 329^[http://aclame.ulb.ac.be/Tools/Prophinder/evaluations_table.html] prophages in the dataset, filtered down to 287 after removing those with fewer than 7 genes; they report a sensitivity and positive predicted value (PPV) of .78 and .95, respectively.  The ProphET paper claims they use the same dataset, although they calculate a sensitivity and PPV of .72 and .75 with Prophinder.  

In conclusion:

- Casjens 2003 apparently describes either 78, 54, or 47 organisms
- ProphET and Prophinder report Prophinder's seensitivity and PPV very differently
- Neither ProphET nor Prophinder release the dataset that they actually use, instead relying on another tool's^[Phage_finder] supplementary table that they filter according to Casjens table
- Maybe if the original supplementary table hadn't been stored in a word doc, some of this confusion could have been avoided.

So, take any tool comparison with large lumpy grain of salt, all because people don't release their evaluation data.



## Merging Data and scoring regiongs 

First, we grouped the happiee results with the correct sequence, using a clunky manner of nesting hits into a single column.  while this is far from ideal, IRanges worked better with this approach, as it struggles with certain operations involving missing data.  We then identified potential hits that will be later filtered.
```{r}
#merged
mdat_pre <- left_join(x=new_dat, y=nested_happie, by=c("short_name", "sequence"))

# make columns with the any hits for prophet or dimob, indexing into the nested coords
mdat <- mdat_pre %>%
  filter(!is.na(valid_locus_start)) %>%
  #mutate(length=valid_locus_end-valid_locus_start) %>%
  group_by(short_name, sequence, valid_locus_start) %>%
  mutate(
    unfiltered_prophet_hit = list(countOverlaps(
      IRanges(prophet[[1]]$start, prophet[[1]]$stop),
      IRanges(valid_locus_start, valid_locus_end),
      type="any", minoverlap = 1)),
    unfiltered_dimob_hit = list(countOverlaps(
      IRanges(dimob[[1]]$start, dimob[[1]]$stop),
      IRanges(valid_locus_start, valid_locus_end),
      type="any", minoverlap = 1)))
# add back in the rows without a start coord (ie, the rows for sequences with no known phage)
missing <- mdat_pre %>% 
  filter(is.na(valid_locus_start)) %>%
  mutate(
    unfiltered_prophet_hit = NA,
    unfiltered_dimob_hit = NA)

mdat <- mdat %>% bind_rows(., missing)
  

# get those hits out of the nested results
for (i in 1:nrow(mdat)){
  dimob_vec <- as.logical(mdat[i, "unfiltered_dimob_hit"][[1]][[1]])
  prophet_vec <- as.logical(mdat[i, "unfiltered_prophet_hit"][[1]][[1]])
  mdat[i, "dimob_start"] <- ifelse(any(dimob_vec),
    mdat[i,"dimob"][[1]] %>% as.data.frame() %>% filter(dimob_vec) %>% head(1) %>% select(start) %>% unlist(), NA)
  mdat[i, "dimob_stop"] <- ifelse(any(dimob_vec),
    mdat[i,"dimob"][[1]] %>% as.data.frame() %>% filter(dimob_vec) %>% head(1) %>% select(stop) %>% unlist(), NA)
  mdat[i, "prophet_start"] <- ifelse(any(prophet_vec),
    mdat[i,"prophet"][[1]] %>% as.data.frame() %>% filter(prophet_vec) %>% head(1) %>% select(start) %>% unlist(), NA)
  mdat[i, "prophet_stop"] <- ifelse(any(prophet_vec),
    mdat[i,"prophet"][[1]] %>% as.data.frame() %>% filter(prophet_vec) %>% head(1) %>% select(stop) %>% unlist(), NA)
}

```

We have the coordinates of possible hits at this point, now lets filter by length, Again, we cant do this ealrier, cause we want to have an overlap over half the length of either the subject or the query, not just one.  This we we could get those these types of hits:

```{bash eval=FALSE}
---------------00000000000---------------- ref
------------oooooooooo-------------------- overlap > 1/2 of ref
----------------ooooo--------------------- overlap < 1/2 ref but > 1/2 query
```

We initally calcualted the detection confusion matrix as follows, counting anything as a positive hit so long as met the above criteria. 
```{r}
mdat <- mdat %>%
  group_by(short_name, sequence, valid_locus_start) %>%
  mutate(
    prophet_hit = ifelse(
      is.na(prophet_start), 
      0,
      countOverlaps(
        IRanges(prophet_start, prophet_stop),
        IRanges(valid_locus_start, valid_locus_end),
        type="any", 
        minoverlap = min(prophet_stop-prophet_start, valid_locus_end-valid_locus_start)/2)),
     dimob_hit = ifelse(
      is.na(dimob_start), 
      0,
      countOverlaps(
        IRanges(dimob_start, dimob_stop),
        IRanges(valid_locus_start, valid_locus_end),
        type="any", 
        minoverlap = min(dimob_stop-dimob_start,        
                         valid_locus_end-valid_locus_start)/2)))
```


Having thoroughly mangled this dataframe, lets get rid of some of those extraneous columns.

```{r}
keepers <- c("short_name","sequence", "status", "valid_locus_start",   "valid_locus_end" ,"prophet_hit", "dimob_hit", "prophet", "enough_genes", "prophet_start", "prophet_stop")
clean_mdat <- mdat%>% select(keepers) 
```


With detection tools, scoring True Negatives comes with a set of caveats, as having a reliable set of such true negatives is often impossible. For prophage detection, this would require having a set of sequences that are of similar length and sequence composition to phages without being an actual prophage. Not being able to calculate true negatives rules out calculating specifity.  As such, looking back at the ProphET paper, they calculalte sensitivity and Positive Predictive Value as follows:


> We  considered  as  true  positives (TP) the overall length  of  the overlap  between  prophage  predictions  and the  gold  standard;false positives (FP) as the overall length of predictions that do not overlap with the gold standard; false negatives (FN) as the overall length of regions of the gold standard that  were  not  included  in  the  predictions  (Supplemental  Figure S3). The  sensitivity (TP / TP + FN)  and  positive  predictive  value  or  PPV  (TP / TP + FP) were computed and compared to those obtained by Prophinder and PHAST.


```{r}

for (s in unique(clean_mdat$sequence)){
  subst <- clean_mdat[clean_mdat$sequence == s,]
  # subst <- clean_mdat[clean_mdat$short_name == clean_mdat$short_name[4], ]
  prophet_hits <- subst$prophet[[1]]
  #if (is.null(prophet_hits))
  if (!is.null(prophet_hits)){
    for (j in 1:nrow(prophet_hits)){
      hit <- prophet_hits[j, ] 
      if (! hit$start %in% subst$prophet_start){
        print("adding row")
        new_row <- subst[1,]
        new_row$status <- "missing"
        new_row$prophet_start <- hit$start
        new_row$prophet_stop <- hit$stop
        new_row$valid_locus_start <- NA
        new_row$valid_locus_end <- NA
        new_row$prophet_hit <- 1
        new_row$enough_genes <- T
        clean_mdat<-rbind(clean_mdat, new_row)
      }
    }
  }
}


# this isn't vectorized cause IRanges doesn't play nicely with vectors involving NAs.
for (i in 1:nrow(clean_mdat)){
  this_row <- clean_mdat[i,]
  if (!is.na(this_row$prophet_start) &
      !is.na(this_row$valid_locus_start) &
      !is.na(this_row$prophet_stop) & 
      !is.na(this_row$valid_locus_end)){
    clean_mdat[i, "assessment"] <- "TP"
    clean_mdat[i, "assessment_range"] <- IRanges::intersect(
        IRanges(this_row$prophet_start, this_row$prophet_stop),
        IRanges(this_row$valid_locus_start, 
                this_row$valid_locus_end))@width
  } else if (!is.na(this_row$prophet_start) &
             is.na(this_row$valid_locus_start)) {
    clean_mdat[i, "assessment"] <- "FP"
    clean_mdat[i, "assessment_range"] <- IRanges(
      this_row$prophet_start, this_row$prophet_stop)@width

  } else if (
    this_row$status != "na" &
    !is.na(this_row$valid_locus_start) &
    is.na(this_row$prophet_start)){
    clean_mdat[i, "assessment"] <- "FN"
    clean_mdat[i, "assessment_range"] <- IRanges(
      this_row$valid_locus_start, this_row$valid_locus_end)@width
  } else {
  clean_mdat[i, "assessment"] <- "other"
  clean_mdat[i, "assessment_range"] <- 0 
  }
}

calc_summary <- function(df){
  TP_sum <- df %>% 
    ungroup() %>% 
    filter(assessment == "TP") %>%
    summarize(sum=sum(assessment_range))

  FP_sum <- df %>% 
    ungroup() %>% 
    filter(assessment == "FP") %>%
    summarize(sum=sum(assessment_range))
  
  FN_sum <- df %>% 
    ungroup() %>% 
    filter(assessment == "FN") %>%
    summarize(sum=sum(assessment_range))
  sensitivity = TP_sum$sum/(TP_sum$sum + FN_sum$sum)
  PPV = TP_sum$sum/(TP_sum$sum + FP_sum$sum)

  print(paste("Sensitivity:", sensitivity))
  print(paste("Positive Predictive Value:", PPV))
}
calc_summary(clean_mdat %>% filter(enough_genes))
```






The "id" column holds the result from happie, with the criteria listed above. 

```{r fig.margin = TRUE}
clean_mdat <- clean_mdat %>% mutate(
  id = ifelse(prophet_hit == 1 & dimob_hit == 1, "dimob,prophet",
             ifelse(prophet_hit == 1, "prophet",
                    ifelse(dimob_hit == 1, "dimob", "na")))
)
ggplot(clean_mdat, aes(x=id, fill=id)) + geom_histogram(stat="count") + coord_flip() + scale_fill_manual(values = c("grey80", "grey60", "grey40", "grey20"), guide=F) + labs(title="Counts of hits by ProphET, dimob, or both") + theme(text=element_text(size=22))

ggplot(clean_mdat %>% filter(enough_genes), aes(x=id, fill=id)) + geom_histogram(stat="count") + coord_flip() + scale_fill_manual(values = c("grey80", "grey60", "grey40", "grey20"), guide=F) + labs(title="Counts of hits by ProphET, dimob, or both", subtitle="Loci with fewer than 7 CDSs were removed") + theme(text=element_text(size=22))


```



We can assess 4 difference scenarios

1. genomes with no phage, no phage detected
2. genomes with no phage, phage detected
3. genomes with phage, no phage detected
4. genomes with phage, phage detected


Dispite having assess which loci consist of fewer than 7 CDSs, I realize I needed to have a column to add whether or not to include a row of the Casjens 2003 data.  Theres a bunch of messy rows (single genes, or very unclear notes), so I went back and added a column to address this, based on the information provided in the description of the locus.

the column has the following identifiers

- intact:  metadata says its excisable, inducible, or filimentous
- intactish: usually described as X-like, meaning its similar to known family of prophage
- partial: a phagey looking operon, but likely missing genes needed for function
- fragments: one or more phage gene
- remnants: degraded, but looks like an old phage
- plasmid: see Borellia genome; the authors thought plasmids would be good to include for that one
- na: data missing altogether, either could access the genome or couldnt relate given locus tags to it.
- unknown: no annotation or annotation unclear
- missing: phages identified with happie but not in original work

```{r  fig.margin = TRUE, warning=FALSE}
ggplot(clean_mdat, aes(x=status, fill=status)) + geom_histogram(stat="count") + coord_flip() +
  labs(title="Phage Metadata") + scale_fill_discrete(guide=F) + theme(text=element_text(size=22))
```


Next, I made a rough "result" column, reporting a conclusion for each entry.
```{r}

valid_hit_categories <-  c("remnants", "partial", "intact", "intactish")
valid_hit_categories <- unique(clean_mdat$status)[!unique(clean_mdat$status)  %in%c(NA, "na", "missing")]
clean_mdat <- clean_mdat %>% group_by(short_name) %>%
  mutate(
    result = ifelse(
      (status %in% valid_hit_categories & !id %in% c("na")) , 
      "True Positives",
      ifelse(
        (status %in% c("na") & id %in% c("na")),
        "'True Negatives'",
        ifelse(
          (status %in% valid_hit_categories & id %in% c("na")),
          "False Negatives",
          ifelse(
            status %in% c("missing"),
            "False Positives / New Phages",
            "Unclear")
        )
      )
    )
  ) 
clean_mdat %>% filter(enough_genes) %>% group_by(result) %>% summarise(n=n())
```


```{r}

#ggplot(new_dat, aes(x=result, fill=result)) + geom_histogram(stat="count") + coord_flip() +scale_fill_discrete(guide=F)
ggplot(clean_mdat %>% filter(enough_genes), aes(x=result, fill=id)) + geom_histogram(stat="count") + coord_flip() + scale_fill_manual(values = c("red", "orange", "grey80", "yellow"))

```



## So whats the story with those "False Positives"?

We were left with a about 11 genomes that ProphET detected as phages that were not reported in Casjens 2013.

```{r}
clean_mdat %>% filter(status == "missing") %>% select(short_name, sequence,  prophet_start, prophet_stop)
```

Based on cursory annotation analysis, I would score these as follows:

Phage:

- Listeria_innocua 2430512 - 2446744
- Salmonella_enterica_CT18 78071 -105813
- Salmonella_enterica_CT18   1920890 - 1937407
- Shigella_flexneri_2a_301  1092698 - 1103027
- Xylella_fastidiosa_Temecula  1194756 - 1217907
-  Xylella_fastidiosa_Temecula  1376468      1392763

Not phage:

- Clostridium acetobutylicum 3420036 - 3439785

Unclear:

- Bifidobacterium longum 1146734 - 1155184
- Salmonella_enterica_CT18  719 - 19889

Pathogenicity Island:

- Staphylococcus_aureus_N315  1862786 - 1881935
- Staphylococcus_aureus_Mu50 1940579 - 1959728

## Conclusions

Together, we can see that ProphET successfully identifies that majority of the phage regions mentioned in Casjens 2003.  Additionally, it identifies regions missing from that list;  given how conservative ProphET seemed in the previous comparison against pther phage ID tools, these regions are of interest, and may be true phages that are more easily identified now in 2019 with all the advances in viral phylogenitcs.  2003 was pre-HTS metagenomics era, and the first marine "metagenome" had just been sequenced (Brietbart 2002).  Research efforts like those from Roux, 2015 expanded the known viral diversity 10-fold (<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4533152/>).  Given that ProphET uses a database updated on a rolling basis means that it it can take advantage of these recent findings.

Assessing the performance of DIMOB alone would not yeild comparable results, as the tool is not strictly a phage detection tool.  Howver we can see that using it in conjunction with ProphET increases the PPV.  What this metric hides, however, are the DIMOB hits that are not phage-related. Because we do not have a good ground-truth dataset for genomic islands.  More information about DIMOB can be found in their 2018 publication^[<https://academic.oup.com/bioinformatics/article/34/13/2161/4904263>]




# Addendum: Assessing plasmid content in E. coli.

The plasmid-finding program used by happie (mlplasmids <https://gitlab.com/sirarredondo/mlplasmids>) uses a taxa-specific support vector machine classifier, and currently has them for E. coli,  Enterococcus faecium, and Klebsiella pneumoniae.  As such, we can only assess how it performs against to E. coli genomes in this dataset, as it is not the appropriate tool to assess plasmid content in diverse taxa.

## Results

| Strain | Sequence      | True    | mlplasmids |  Note |
|--------|---------------|---------|------------|-------|
| K-12   | NC_000913.3   | Chrom   | Chrom      |       |
| CFT073 | NC_004431.1   | Chrim   | Chrom      |       |
| EDL933 | NZ_CP008957.1 | Chrom   | Chrom      | X      |
| EDL933 | NZ_CP008958.1 | Plasmid | Plasmid    | X     |

Note that EDL933 was excluded for reasons above, but can still be assessed for plasmid content.

This doesn't show a ton of interest, but doing a whole analysis to test how well mlplasmids works on E coli genomes would just be recapitulating their publication, and is outside the scope of this validation.



